File: go-tiny-claw/internal/feishu/bot.go
// internal/feishu/bot.go
package feishu

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"strings"

	larkcore "github.com/larksuite/oapi-sdk-go/v3/core"
	"github.com/larksuite/oapi-sdk-go/v3/event/dispatcher"
	larkim "github.com/larksuite/oapi-sdk-go/v3/service/im/v1"
	larkws "github.com/larksuite/oapi-sdk-go/v3/ws"
	"github.com/yourname/go-tiny-claw/internal/engine"

	lark "github.com/larksuite/oapi-sdk-go/v3"
)

// FeishuBot 封装了飞书机器人的配置与核心业务流
type FeishuBot struct {
	client    *lark.Client
	appID     string
	appSecret string
	engine    *engine.AgentEngine // 持有核心引擎引用
}

func NewFeishuBot(appID, appSecret string, eng *engine.AgentEngine) *FeishuBot {
	// 实例化飞书官方客户端
	client := lark.NewClient(appID, appSecret)

	return &FeishuBot{
		client:    client,
		appID:     appID,
		appSecret: appSecret,
		engine:    eng,
	}
}

// Start 通过长连接（WebSocket）方式启动飞书事件监听，无需公网 IP 和内网穿透
func (b *FeishuBot) Start(ctx context.Context) error {
	// 长连接模式下，verifyToken 和 encryptKey 必须传空字符串
	eventHandler := dispatcher.NewEventDispatcher("", "").
		OnP2MessageReceiveV1(func(ctx context.Context, event *larkim.P2MessageReceiveV1) error {
			contentStr := *event.Event.Message.Content
			contentStr = strings.TrimPrefix(contentStr, `{"text":"`)
			contentStr = strings.TrimSuffix(contentStr, `"}`)

			chatId := *event.Event.Message.ChatId
			log.Printf("[Feishu] 收到会话 %s 消息: %s\n", chatId, contentStr)

			go b.handleAgentRun(chatId, contentStr)

			return nil
		}).
		OnP2MessageReadV1(func(ctx context.Context, event *larkim.P2MessageReadV1) error {
			return nil
		})

	// 创建长连接客户端
	cli := larkws.NewClient(b.appID, b.appSecret,
		larkws.WithEventHandler(eventHandler),
		larkws.WithLogLevel(larkcore.LogLevelDebug),
	)

	log.Println("🚀 go-tiny-claw 正在通过长连接（WebSocket）方式连接飞书...")

	// Start 会阻塞主线程，直到连接断开
	return cli.Start(ctx)
}

// handleAgentRun 是连接飞书与底层引擎的桥梁
func (b *FeishuBot) handleAgentRun(chatId string, prompt string) {
	// 为当前聊天窗口实例化一个专属的 Reporter
	reporter := &FeishuReporter{
		client: b.client,
		chatId: chatId,
	}

	// 启动引擎！
	err := b.engine.Run(context.Background(), prompt, reporter)
	if err != nil {
		reporter.sendMsg(fmt.Sprintf("❌ Agent 运行崩溃: %v", err))
	}
}

// ==========================================
// FeishuReporter: 将引擎的输出格式化后发给飞书
// ==========================================
type FeishuReporter struct {
	client *lark.Client
	chatId string
}

// sendMsg 封装了调用飞书 OpenAPI 发送卡片/文本的操作
func (r *FeishuReporter) sendMsg(text string) {
	// 构建文本消息内容
	textContent := map[string]string{
		"text": text,
	}
	contentBytes, _ := json.Marshal(textContent)
	contentStr := string(contentBytes)

	msgReq := larkim.NewCreateMessageReqBuilder().
		ReceiveIdType("chat_id").
		Body(larkim.NewCreateMessageReqBodyBuilder().
			ReceiveId(r.chatId).
			MsgType("text").
			Content(contentStr).
			Build()).
		Build()

	_, _ = r.client.Im.Message.Create(context.Background(), msgReq)
}

func (r *FeishuReporter) OnThinking(ctx context.Context) {
	// 仅发一个轻量级提示，避免飞书刷屏
	r.sendMsg("🤔 模型正在慢思考 (Thinking)...")
}

func (r *FeishuReporter) OnToolCall(ctx context.Context, toolName string, args string) {
	r.sendMsg(fmt.Sprintf("🛠️ **正在执行工具**：`%s`\n参数：`%s`", toolName, args))
}

func (r *FeishuReporter) OnToolResult(ctx context.Context, toolName string, result string, isError bool) {
	if isError {
		r.sendMsg(fmt.Sprintf("⚠️ **执行报错** (%s)：\n%s", toolName, result))
	} else {
		// 成功时仅汇报成功，不刷全量日志
		r.sendMsg(fmt.Sprintf("✅ **执行成功** (%s)", toolName))
	}
}

func (r *FeishuReporter) OnMessage(ctx context.Context, content string) {
	// 将模型最终的纯文本回答发给用户
	r.sendMsg(content)
}

// 编译时类型检查：确保 FeishuReporter 实现了 Reporter 接口
var _ engine.Reporter = (*FeishuReporter)(nil)

File: python-tiny-claw/internal/feishu/bot.py
from __future__ import annotations
import json
import logging
import os
import threading
from typing import Any

from internal.engine import Reporter, AgentEngine
from internal.context import Session
from internal.schema import Message, Role
from .approval import global_approval_mgr


class FeishuReporter(Reporter):
    def __init__(self, client: Any, chat_id: str):
        self.client = client
        self.chat_id = chat_id

    def send_msg(self, text: str) -> None:
        try:
            content = json.dumps({"text": text})
            logging.info(f"[Feishu] 发送消息到 {self.chat_id}: {text[:100]}...")
        except Exception as e:
            logging.error(f"[Feishu] 发送消息失败: {e}")

    def on_thinking(self) -> None:
        self.send_msg("🤔 模型正在慢思考 (Thinking)...")

    def on_tool_call(self, tool_name: str, args: str) -> None:
        self.send_msg(f"🛠️ **正在执行工具**：`{tool_name}`\n参数：`{args}`")

    def on_tool_result(self, tool_name: str, result: str, is_error: bool) -> None:
        if is_error:
            self.send_msg(f"⚠️ **执行报错** ({tool_name})：\n{result}")
        else:
            self.send_msg(f"✅ **执行成功** ({tool_name})")

    def on_message(self, content: str) -> None:
        self.send_msg(content)


class FeishuBot:
    def __init__(self, engine: AgentEngine, sess: Session):
        self.app_id = os.getenv("FEISHU_APP_ID", "")
        self.app_secret = os.getenv("FEISHU_APP_SECRET", "")
        self.engine = engine
        self.sess = sess
        self.reporter = None

        if not self.app_id or not self.app_secret:
            logging.warning("未配置 FEISHU_APP_ID/FEISHU_APP_SECRET，飞书机器人功能不可用")

    def _handle_message(self, chat_id: str, content_str: str) -> None:
        logging.info(f"[Feishu] 收到会话 {chat_id} 消息: {content_str}")

        if content_str.startswith("approve "):
            task_id = content_str[len("approve "):].strip()
            global_approval_mgr.resolve_approval(task_id, True, "人类管理员已批准操作")
            logging.info(f"[Feishu] 会话 {chat_id}: ✅ 已为您批准任务 {task_id}")
            return

        if content_str.startswith("reject "):
            task_id = content_str[len("reject "):].strip()
            global_approval_mgr.resolve_approval(task_id, False, "人类管理员认为该操作存在极高风险，已无情拒绝")
            logging.info(f"[Feishu] 会话 {chat_id}: 🚫 已拒绝任务 {task_id}")
            return

        threading.Thread(target=self._handle_agent_run, args=(chat_id, content_str), daemon=True).start()

    def _handle_agent_run(self, chat_id: str, prompt: str) -> None:
        reporter = FeishuReporter(None, chat_id)
        self.reporter = reporter
        self.sess.append(Message(role=Role.USER, content=prompt))
        try:
            self.engine.run(self.sess, reporter)
        except Exception as e:
            reporter.send_msg(f"❌ Agent 运行崩溃: {str(e)}")

    def start(self) -> None:
        logging.info("飞书机器人启动占位实现 - 需要配置 lark-oapi SDK")
